### ISSUE TITLE
Setting up PMD for Salesforce project in Neovim

### ISSUE THREAD

**Issue by ognjen-vuceljic** [xixiaofinland/dotfiles-nix#1]
Hello!

I'm not sure if this is the best place to post such a question, but looking at dotfiles and neovim config, I tried replicating some of the items in order to add PMD to my Salesforce project.

I looked at [this discussion](https://www.reddit.com/r/neovim/comments/18hclzu/pmd_v7_code_analayzer_not_working_in_nullls_or/) and tried applying the mentioned fixes, but had no luck.

Once I added the following part to `none-ls.lua` plugin

```
null_ls.builtins.diagnostics.pmd.with({
          filetypes = { "apex" },
          args = { "check", "--dir", "$ROOT", "--format", "json" },
          extra_args = { "--rulesets", "apex_ruleset.xml", },
        }),
```

the following error was received

<img width="1416" alt="image" src="https://github.com/xixiaofinland/dotfiles-nix/assets/102177460/bf951d78-56fb-444e-9e90-1fa33cc4175a">

Could you, please, let me know what the configuration issue might be?

Appreciate the help!

@xixiaofinland

**Comment by xixiaofinland** [comment]
Do you use PMD v6 or v7? If you use v6, it should be straightforward.

If you use v7, none-ls looks for a "pmd" executable, which corresponds to "pmd check" in v7.
To address this, you can create a bash script and place it in your $PATH for none-ls to invoke.

To test the result, when you run "pmd" manually in the console now, you will actually be executing this bash script, which runs "pmd check" as a wrapper and attach your inputs.

```bash
#!/usr/bin/env bash
path/to/pmd/bin/pmd check "$@"
``

**Comment by xixiaofinland** [comment]
I'm currently using v6, but this setting worked for me for v7.

https://github.com/xixiaofinland/dotfiles-nix/blob/4ec92187c21e018772ef391577146e483653f51e/dotfiles/nvim/lua/plugins/null_ls.lua#L21

**Comment by ognjen-vuceljic** [comment]
Thank you @xixiaofinland 

I am almost there, but a few final steps are missing.

Apologies, I didn't understand the following part:

`If you use v7, none-ls looks for a "pmd" executable, which corresponds to "pmd check" in v7.
To address this, you can create a bash script and place it in your $PATH for none-ls to invoke.`

- what should be the name of that script (can it be `pmd.sh`, or it has to be somewhere else)?
- my $PATH config appears to be pretty long --> any particular place where i need to put this? For example, the $PATH lists the hierarchy of different folders, so I am not quite sure where to put the script

Basically, the part I didn't catch was where exactly, and how, to create the script. My understanding is that it needs to be a `.sh` file with the content from message above, and to be put in a particular folder.

Btw, this is what I am getting at the moment

![image](https://github.com/xixiaofinland/dotfiles-nix/assets/102177460/277c4b78-fbba-4144-82d5-283b7696c0fd)

PMD may have even run against this file, but no diagnostic messages are shown.

Many thanks for the help!

**Comment by xixiaofinland** [comment]
you need to name the bashscript as `pmd` without suffix, and put it into one of your $PATH directories, [explanation](https://askubuntu.com/questions/551990/what-does-path-mean).

By doing this, in console you can call `pmd` from anywhere, just like other globally installed tools `node`, `npm` etc.

**Comment by xixiaofinland** [comment]
I suggest you understand what $PATH does first. None-ls is simply invoking `pmd` with parameters defined in the configuration.

If you can run the same commands in console successfully, none-ls should be ok too. If no, then dig a bit deeper.

**Comment by ognjen-vuceljic** [comment]
Thank you @xixiaofinland , will take a look.

I think we can close this issue!
